dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -418,24 +375,124 @@ private Map<String, Object> convertToolToSingleStepWorkflow(Map<String, Object>
+        map.computeIfPresent("requirements", (k, v) -> convertToList(v, "class", where));
+        return map;
     }
 
     /**
-     * Extract the workflow step ID from the full normalized ID that cwljava returns, then prepend the parent ID if it is not null.
+     * Parse a map to an instance of the specified class.
+     * @param obj the object to parse
+     * @param klass the class to parse to
+     * @param description description of the object we are parsing, for error reporting purposes
      */
-    private String convertStepId(String fullStepId, String parentStepId) {
-        String[] parts = fullStepId.split("/");
-        String thisStepId = parts[parts.length - 1];
-        return parentStepId == null ? thisStepId : parentStepId + "." + thisStepId;
+    private <T> T parseWithClass(Object obj, Class<T> klass, String description) {
+        if (obj instanceof Map) {
+            Gson gson = CWL.getTypeSafeCWLToolDocument();
+            Map<Object, Object> map = convertRequirementsAndHintsToLists((Map<Object, Object>)obj, "in the requirements/hints of " + description);
+            return gson.fromJson(gson.toJson(map), klass);
+        } else {
+            String message = CWL_PARSE_ERROR + "malformed cwl in " + description;
+            LOG.error(message);
+            throw new CustomWebApplicationException(message, HttpStatus.SC_UNPROCESSABLE_ENTITY);
+        }
+    }
+
+    private Workflow parseWorkflow(Object workflowObj) {
+        return parseWithClass(workflowObj, Workflow.class, "workflow");
+    }
+
+    private CommandLineTool parseTool(Object toolObj) {
+        return parseWithClass(toolObj, CommandLineTool.class, "tool");
+    }
+
+    private ExpressionTool parseExpressionTool(Object expressionToolObj) {
+        return parseWithClass(expressionToolObj, ExpressionTool.class, "expression tool");
+    }
+
+    private WorkflowStep parseWorkflowStep(Object workflowStepObj) {
+        return parseWithClass(workflowStepObj, WorkflowStep.class, "workflow step");
+    }
+
+    private boolean isProcessWithClass(Object candidate, String classValue) {
+        return candidate instanceof Map && classValue.equals(((Map<Object, Object>)candidate).get("class"));
+    }
+
+    private boolean isWorkflow(Object candidate) {
+        return isProcessWithClass(candidate, WORKFLOW);
+    }
+
+    private boolean isTool(Object candidate) {
+        return isProcessWithClass(candidate, COMMAND_LINE_TOOL);
+    }
+
+    private boolean isExpressionTool(Object candidate) {
+        return isProcessWithClass(candidate, EXPRESSION_TOOL);
+    }
+
+    private boolean isOperation(Object candidate) {
+        return isProcessWithClass(candidate, OPERATION);
+    }
+
+    private String getWorkflowStepId(WorkflowStep workflowStep) {
+        CharSequence id = workflowStep.getId();
+        if (id == null) {
+            String message = CWL_PARSE_ERROR + "missing id in workflow step";
```

Comment:
Is it possible to provide a more specific error message? I guess you can't say "Missing id in workflow step with id \<id\>" since you don't have an id :), but if there were anything that could point the user more closely to the problem (it may not be possible)....

Reply:
I don't believe we currently surface this error message to the user (because if the DAG/tool generation fails, it does so silently).  So, until that changes, I'm gonna suggest that it's good enough...